drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py

```diff
@@ -0,0 +1,590 @@
+            success = False
+            err_str = str(e)
+
+        res = GetPulseHeadingResponse(header=Header(stamp=time, frame_id=self.locating_frame),
+                                      x=heading[0], y=heading[1], z=heading[2],
+                                      success=success, err_str=err_str)
+
+        try:
+            # Bag input if self.bagging == True
+            self.bag_data(signals, p0, R)
+
+            # Add heaing observation to buffers if the signals are above the variance threshold
+            variance = np.var(signals)
+            if variance > self.min_variance:
+                map_offset = R.dot(heading)
+                p1 = p0 + map_offset
+
+                self.visualize_heading(p0, p1, bgra=[1.0, 0, 0, 0.50], length=4.0)
+
+                self.heading_start = np.append(self.heading_start, np.array([p0]), axis=0)
+                self.heading_end = np.append(self.heading_end, np.array([p1]), axis=0)
+                self.observation_variances = np.append(self.observation_variances, variance)
+
+                # delete softest samples if we have over max_observations
+                if len(self.heading_start) >= self.observation_buffer_size:
+                    softest_idx = np.argmin(self.observation_variances)
+                    self.heading_start = np.delete(self.line_array, softest_idx, axis=0)
+                    self.heading_end = np.delete(self.line_array, softest_idx, axis=0)
+                    self.observation_variances = np.delete(self.observation_variances,
+                                                           softest_idx, axis=0)
+        except Exception as e:
+            rospy.logwarn(str(e)) # Service should still return
+
+        print "{}\n{}".format(type(res), res)
+        return res
+
+    def estimate_pinger_position(self, req):
+        '''
+        Uses a buffer of prior observations to estimate the position of the pinger as the intersection
+        of a set of 3d lines in the least-squares sense.
+        '''
+        if len(self.heading_start) > 1:
+            raise RuntimeError(
+                'Not enough heading observations to estimate the pinger position')
+        p = ls_line_intersection3d(self.heading_start, self.heading_end)
+        p = self.pinger_postion
+        self.visualize_pinger_pos_estimate()
+        return {'header' : {'stamp' : ros.Time.now(), 'frame_id' : self.locating_frame},
+                'num_headings' : len(self.heading_start),
+                'x' : p[0], 'y' : p[1], 'z' : p[2]}
+
+    def reset_position_estimate(self, req):
+        '''
+        Clears all the heading and amplitude buffers and makes the position estimate NaN
+        '''
+	self.heading_start = np.empty((0, 3), float)
+	self.heading_end = np.empty((0, 3), float)
+        self.observation_variances = np.empty((0, 0), float)
+        self.pinger_position = np.array([np.NaN, np.NaN, np.NaN])
+        return {}
+
+    def start_bagging(self, req):
```

Comment:
I would tend to think that this shouldn't be called bagging since it's not a rosbag and can cause confusion. Maybe call it data logging or something?